### PR TITLE
Policy: Allowed + Constraints + Selector

### PR DESCRIPTION
**Pull request by JoshVanL**
This PR adds a number of things.

I have split out `spec.issuerRefSelector` to become `spec.selector.issuerRef`.

The old `attributes` approver has been split up into `allowed` and `constraints`.

Fields under `allowed` which include things like `dnsNames`, `commonName` are denied by default. This means the things listed under `allowed` effectively give the certificate holder _more_ privileges so a user likely wants to list everything an identity should have.

Fields under `constraints` include things like `maxDuration` or `keyAlgorithm` and are allow all by default. While these fields don't give the certificate holder more privileges,  they are things that some users may be interested in limiting.

I have expanded the controller BDD tests quite a bit so we should have much better coverage now. If you think of something I have missed please say :smile: 

/assign @irbekrm 
/hold

### CODE REVIEW

**Comment by JoshVanL** [comment]
/

File: pkg/internal/controllers/test/ready.go
Comment (irbekrm) [inline pkg/internal/controllers/test/ready.go:1]: Question- wouldn't it be better to name these files `_test.go` so they don't get included in the built binary?

File: pkg/internal/controllers/test/review.go
Comment (irbekrm) [inline pkg/internal/controllers/test/review.go:1]: The tests seem really thorough to me 👍🏼 
Minor suggestion: perhaps some test cases could also include constraints?

File: pkg/approver/constraints/validation.go
Comment (irbekrm) [inline pkg/approver/constraints/validation.go:69]: Question: would there be any reason for a constraint with a 0 max key size?

**Review by irbekrm — APPROVED** [review]
I've read through the code and the approach makes sense to me.

I guess the one potential gothca I could imagine was if someone thought that they could set constraints on one policy (perhaps some cluster-wide constraints?) and allowed`s on another policy- since the policies are not additive, this would just result in the constraints essentially having no effect. Would it be worth to mention/highlight this somewhere? Or, perhaps not since that seems to be a wrong way to think about these policies altogether?

Otherwise I've added a couple nits, I see that the PR is currently held, so feel free to address or unhold 👍🏼 

/lgtm

File: pkg/approver/allowed/evaluator.go
Comment (JoshVanL) [reply pkg/approver/allowed/evaluator.go:81]: Yes, it would be better if you could define subnets in this way. This would likely be another allowed field I think (`ipSubnets`), but this can be done in a followup PR.

File: pkg/approver/allowed/evaluator.go
Comment (JoshVanL) [reply pkg/approver/allowed/evaluator.go:106]: Correct. If an entity is allowed to request isCA==true, we don't care if they request a cert which is isCA==false

File: pkg/approver/allowed/evaluator.go
Comment (JoshVanL) [reply pkg/approver/allowed/evaluator.go:114]: So this is a signer concern on _what_ they do with given usages (ignore them, change them, etc. etc.). Approver is only concerned with what's on the actual request.

It should be a follow-up PR, but we should think about how we are going to add evaluations for extensions in the actual; CSR document https://github.com/cert-manager/istio-csr/blob/master/pkg/server/internal/extensions/extensions.go

File: pkg/approver/allowed/evaluator_test.go
Comment (JoshVanL) [reply pkg/approver/allowed/evaluator_test.go:265]: Thanks :slightly_smiling_face: 

I have always thought about this behavior but didn't know there was a stdlib func for it.

File: pkg/internal/controllers/test/ready.go
Comment (JoshVanL) [reply pkg/internal/controllers/test/ready.go:1]: They will not get put into the built binary because they are not referenced by the program.

Non-referenced func and code is omitted from built Go programs

File: pkg/approver/constraints/validation.go
Comment (JoshVanL) [reply pkg/approver/constraints/validation.go:69]: Yes, this makes a robust "deny-all" policy possible.https://github.com/cert-manager/approver-policy/blob/bde0903a2c66719651c41c4348207ff7621ddf15/docs/examples/deny-all.yaml

**Comment by JoshVanL** [comment]
Thanks @irbekrm, I have responded to your comments and implemented the suggestions.

Hopefully #31 will help with giving proper docs on what the behavior is to users.

**Comment by irbekrm** [comment]
Looks good to me- thanks for making the changes and answering all the questions

/hold cancel

File: pkg/approver/allowed/evaluator.go
Comment (irbekrm) [reply pkg/approver/allowed/evaluator.go:81]: I imagine the range field could just replace this one (you could specify a single IP with `/32` as for various firewall configs etc), so we don't have two fields for the same thing.

**Pull request merged**
